### ISSUE TITLE
inst_label

### ISSUE THREAD

**Issue by Akihiko0123** [mahmoodlab/CLAM#135]
Thank you for sharing this amazing method.
I would appreciate it if you could let me know on the followings.

In the method "forward" of CLAM_SB, inst_label seems to be switched according to the label of the slide.
In a for loop "for i in range(len(self.instance_classifiers))", 
if i=0 and the slide label is "normal_tissue", inst_label become 1 and calculated in self.inst_eval.
if i=1 and the slide label is "tumor_tissue", inst_label become 1 and calculated in self.inst_eval.
(especially in case self.subtyping==FALSE)
Since both are calculated samely in self.inst_eval, I am not sure what is the difference between these two.
Are these different because different networks(classifier) which learn weight differently are used for each class?

Thank you.
```
            for i in range(len(self.instance_classifiers)):
                inst_label = inst_labels[i].item()
                classifier = self.instance_classifiers[i]
                if inst_label == 1: #in-the-class:
                    instance_loss, preds, targets = self.inst_eval(A, h, classifier)
                    all_preds.extend(preds.cpu().numpy())
                    all_targets.extend(targets.cpu().numpy())
                else: #out-of-the-class
                    if self.subtyping:
                        instance_loss, preds, targets = self.inst_eval_out(A, h, classifier)
                        all_preds.extend(preds.cpu().numpy())
                        all_targets.extend(targets.cpu().numpy())
                    else:
                        continue
```

**Comment by fedshyvana** [comment]
The instance-level classifiers are one vs. rest and you have one classifier per class. So for each WSI, depending on the the WSI-level target, the corresponding instance-level classifier is used/trained to predict the high attention patches as positive, and the remaining instance-level classifiers are used/trained to predict them as negative. Granted if you only consider the binary tumor vs. normal setting, having 2 one vs. rest classifiers seem a bit redundant, but i suppose its the most general formulation because even for 2-class problems other than tumor vs. normal (e.g. disease class A vs. disease class B), there could technically be more than 2 classes at the patch-level (there's patches of class A, of class B, and of neither, such as normal tissue/debris).

**Comment by Akihiko0123** [comment]
Thank you so much for letting me know the details.
I understood that having classifiers for both positive and negative would be effective especially in subtyping even 2-classes.

Also, although it might be obvious, in case of tumor vs. normal, probably the weight would be different between each classifiers after study even if it seems to be same fully connected layers (512 -> 2).

I would appreciate it if you could tell me if I am misunderstanding something.

**Comment by fedshyvana** [comment]
yes that is correct - weights would be different.

**Comment by Akihiko0123** [comment]
I am relieved to hear that, thank you!

**Comment by LiuJin629** [comment]
I would like to ask the author a question:
Model(data, label=label, instance_eval=True) is used during the model training process, which means that the instance-level is used for patch-level training during the training process, and the label information of the slide is used to determine these high-level Attention and low-attention patches are assigned to the corresponding fully connected layers.
But what puzzles me is that only "model(data)" is used during the testing process. In this case, "instance_eval" is False, then the patch-level fully connected layer trained during the training process is not used in the testing phase. So, what role do the N fully connected layers trained in the training phase play in the testing phase? Or is it just using instance_loss to back propagate and update the parameters in the slide level classifier (i.e. "self.attention_net = nn.Sequential(*fc)" and "self.classifiers = nn.Linear(size[1], n_classes )" parameters)?
Looking forward to your answer very much!
